### ISSUE TITLE
Stop reason: signal SIGSEGV: invalid address (fault address: 0x0)

### ISSUE THREAD

**Issue by sotrh** [gfx-rs/wgpu-rs#194]
I've been working on a model loading tutorial, and I just got to the part where I'm creating `BindGroup`s for each of the model's materials. The code runs fine if I use a bind group created outside of the model loading code. If I try to use a bind group created inside the model loading code it crashes with a segfault on the `draw_indexed` call.

```rust
fn draw_mesh_instanced(&mut self, mesh: &Mesh, material: &Material, instances: Range<u32>) {
    self.set_vertex_buffers(0, &[(&mesh.vertex_buffer, 0)]);
    self.set_index_buffer(&mesh.index_buffer, 0);
    self.set_bind_group(0, &material.bind_group, &[]);
    self.draw_indexed(0..mesh.num_elements, 0, instances);
}
```

I load the models as follows:

```rust


impl Model {
    pub fn load<P: AsRef<Path>>(device: &wgpu::Device, layout: &wgpu::BindGroupLayout, path: P) -> Result<(Self, Vec<wgpu::CommandBuffer>), failure::Error> {
        let (obj_models, obj_materials) = tobj::load_obj(path.as_ref())?;

        // We're assuming that the texture files are stored with the obj file        
        let containing_folder = path.as_ref().parent().unwrap();

        // Our `Texure` struct currently returns a `CommandBuffer` when it's created so we need to collect those and return them.
        let mut command_buffers = Vec::new();

        let mut materials = Vec::new();
        for mat in obj_materials {
            let diffuse_path = mat.diffuse_texture;
            let (diffuse_texture, cmds) = texture::Texture::load(&device, containing_folder.join(diffuse_path))?;
            
            let bind_group = device.create_bind_group(&wgpu::BindGroupDescriptor {
                layout,
                bindings: &[
                    wgpu::Binding {
                        binding: 0,
                        resource: wgpu::BindingResource::TextureView(&diffuse_texture.view),
                    },
                    wgpu::Binding {
                        binding: 1,
                        resource: wgpu::BindingResource::Sampler(&diffuse_texture.sampler),
                    },
                ]
            });

            materials.push(Material {
                name: mat.name,
                diffuse_texture,
                bind_group,
            });
            command_buffers.push(cmds);
        }

        let mut meshes = Vec::new();
        for m in obj_models {
            let mut vertices = Vec::new();
            for i in 0..m.mesh.positions.len() / 3 {
                vertices.push(ModelVertex {
                    position: [
                        m.mesh.positions[i * 3],
                        m.mesh.positions[i * 3 + 1],
                        m.mesh.positions[i * 3 + 2],
                    ],
                    tex_coords: [
                        m.mesh.texcoords[i * 2],
                        m.mesh.texcoords[i * 2 + 1],
                    ],
                    normal: [
                        m.mesh.normals[i * 3],
                        m.mesh.normals[i * 3 + 1],
                        m.mesh.normals[i * 3 + 2],
                    ],
                });
            }

            let vertex_buffer = device
                .create_buffer_mapped(vertices.len(), wgpu::BufferUsage::VERTEX)
                .fill_from_slice(&vertices);

            let index_buffer = device
                .create_buffer_mapped(m.mesh.indices.len(), wgpu::BufferUsage::INDEX)
                .fill_from_slice(&m.mesh.indices);

            meshes.push(Mesh {
                name: m.name,
                vertex_buffer,
                index_buffer,
                num_elements: m.mesh.indices.len() as u32,
                material: m.mesh.material_id.unwrap_or(0),
            });
        }
        
        Ok((Self { meshes, materials, }, command_buffers))
    }
}
```

The tutorial as of today is here:
https://github.com/sotrh/learn-wgpu/tree/model/code/beginner/tutorial9-models

The model loading code is here:
https://github.com/sotrh/learn-wgpu/blob/model/code/beginner/tutorial9-models/src/model.rs

My drivers are up to date. I'm running Linux Mint.

Here's a screenshot of the call stack.

![Screenshot from 2020-03-10 22-12-58](https://user-images.githubusercontent.com/7904656/76381967-b4b1b280-631c-11ea-93ad-8a33d3d0e548.png)

Hopefully I've provided enough information. If not let me know.

**Comment by kvark** [comment]
Fun to see Vulkan validation layers crashing :)
This is just a fix of https://github.com/gfx-rs/wgpu/pull/495 missing from the published 0.4 version. You can work around this by making sure you bind 0 first and then 1, like here - https://github.com/kvark/learn-wgpu/commit/75024dfc1bc6526342b4f45d27d018c0b8fd018f

The good thing is that re-ordering your bindings makes sense regardless of the bug. It's expected to have less-frequently-changing bind groups at lower indices and as data is more specific in needs to raise in binding indices. E.g.. your per-frame uniforms would be in (0), your per pass data would be in (1), your per-instance parameters in (2). So my recommendation would be to just change that order, so that you'd be binding group (0) outside of `draw_mesh`, and the material is bound to group (1), which naturally works around the bug (that is otherwise fixed in `master`).

If that's not convenient, we can make a patch to 0.4. Truth is - 0.5 went very far ahead at this time, so we don't expect many users to stick with 0.4 once it's not the latest.

**Comment by sotrh** [comment]
Yep, that fixed it. That's a bizarre bug. Good to see it's fixed in 0.5
